nemo/collections/nlp/models/language_modeling/megatron_retrival_model.py

```diff
@@ -0,0 +1,537 @@
+                    total_samples=len(dataset),
+                    consumed_samples=consumed_samples,
+                    micro_batch_size=self._cfg.micro_batch_size,
+                    global_batch_size=self._cfg.global_batch_size,
+                    data_parallel_rank=parallel_state.get_data_parallel_rank(),
+                    data_parallel_size=parallel_state.get_data_parallel_world_size(),
+                    drop_last=self._cfg.get('drop_last', True),
+                )
+            elif self._cfg.data.dataloader_type == 'cyclic':
+                batch_sampler = MegatronPretrainingRandomBatchSampler(
+                    total_samples=len(dataset),
+                    consumed_samples=consumed_samples,
+                    micro_batch_size=self._cfg.micro_batch_size,
+                    global_batch_size=self._cffg.global_batch_size,
+                    data_parallel_rank=parallel_state.get_data_parallel_rank(),
+                    data_parallel_size=parallel_state.get_data_parallel_world_size(),
+                    drop_last=self._cfg.get('drop_last', True),
+                )
+            else:
+                raise Exception(f'{self._cfg.dataloader_type} dataloader type is not supported.')
+        else:
+            raise ValueError('cfg.data.dataloader_type not found. Must be "single" or "cyclic"')
+
+        # Torch dataloader.
+        return torch.utils.data.DataLoader(
+            dataset, batch_sampler=batch_sampler, num_workers=self._cfg.data.num_workers, pin_memory=True,
+        )
+
+    def setup(self, stage=None):
+        resume_checkpoint_path = self.trainer._checkpoint_connector.resume_from_checkpoint_fit_path
+        if resume_checkpoint_path:
+            try:
+                init_consumed_samples = int(
+                    float(re.findall(r"consumed_samples\=([0-9]+.[0-9]+)", resume_checkpoint_path)[0])
+                )
+            except (ValueError, TypeError):
+                logging.warning("Cannot parse the checkpoint file to get the consumed samples. assume it is zero.")
+                init_consumed_samples = 0
+        else:
+            init_consumed_samples = 0
+        self.init_consumed_samples = init_consumed_samples
+
+        """A PTL method to setup the training, validation and test datasets."""
+        if stage == 'predict':
+            return
+        if self._train_dl is not None and self._validation_dl is not None:
+            return
+        self.build_train_valid_test_datasets()
+        self.setup_training_data(self._cfg.data)
+        self.setup_validation_data(self._cfg.data)
+        self.setup_test_data(self._cfg.data)
+
+        # when using pipeline model parallel the final stage need to initialize word embeddings
+        if parallel_state.get_pipeline_model_parallel_world_size() > 1:
+            self.enc_dec_model.sync_initial_word_embeddings()
+            self.enc_dec_model.sync_initial_position_embeddings()
```

Comment:
Probably better to comment this out until we support PP

Reply:
cleaned up